src/hardware/sblaster.cpp

```diff
@@ -2816,8 +2903,52 @@ static uint8_t read_sb(const io_port_t port, const io_width_t)
 	return 0xff;
 }
 
+static void generate_frame();
+
+static void sblaster_pic_callback([[maybe_unused]] const uint32_t)
+{
+	assert(sblaster);
+	if (!sblaster->channel->is_enabled) {
+		return;
+	}
+
+	// Generate a new frame if we no longer have a balance.
+	if (!queued_frames_balance) {
+		generate_frame();
+	}
+
+	// This callback is fired at a rate of once every channel frame, so we
+	// decrement the positive balance because we've consumed one frame
+	// simply through the passage of time.
+	if (queued_frames_balance) {
+		--queued_frames_balance;
+	}
+
+	PIC_AddEvent(sblaster_pic_callback, sblaster->channel->GetMsPerFrame());
```

Comment:
Not sure if I'm a fan of this.  Maybe I could be convinced if you explained your rational.  This is adding a lot of events doing this for every single audio frame.  The PIC queue is implemented as a linked list so it's not efficient for large amounts of things.

I'm also not seeing what the advantage is over using a loop like I had before.  From my understanding, the PIC queue fires every 1ms after the CPU emulation is done.  So these events will either happen immediately following this or on the next tick.  It should work the same as a loop but with extra steps and "PIC abuse" as @johnnovak likes to call it.

Reply:
General comment: simpler (and I daresay, more primitive) solutions are better that get the job done, and no more PIC abuse please if it can be avoided 😆

Reply:
The way DMA and the SB work together is that the card requests bytes from the DMA controller, sample by sample across time, that the card decodes and then plays.
 
The SB doesn't have a buffer to pull in multiple samples at a time or buffer block locally.. instead the SB literally treats the DMA memory as it's own, and it just reads from it at the playback rate. That's how CDreams works... it continually overwrites the DMA memory (instead of setting up big chunky transfers). 

So that's how the current PR gets this and other demos right, but also works fine for quake and other DMA games.

But your right @weirddan455 - running the PIC callback at a per-frame timing is more expensive than running the callback once every tick; (I could measure between a 2% and 5% fps loss in quake in the per-frame scenario)

So I'm adding a per-tick approach (N frames per 1 tick), exactly like you've done in the GUS and other cards. This mode will be used when the DMA transfer uses a larger multi-frame count.

Reply:
Ok, that's fine, then I guess adding a secondary emulator-only PIC to handle extra internal emulator-only duties that the real PIC was never intended to do is a job for another day.

Reply:
>  The SB doesn't have a buffer to pull in multiple samples; instead it treats the DMA memory as its own and reads from it at the playback rate.
>
>  (Crystal Dreams continually overwrites the DMA memory sample by sample; if the SB read in chunks, instead of setting up big chunky transfers).

Correct, and that's the same problem I ran into with GUS on #3993  I haven't researched it heavily yet but my theory is most games are going to treat DMA sound as a ring buffer and write ahead of where the SB is reading from (but also making sure not to write so far ahead the sound overlaps the read position).

Performance is a concern but also the ability to debug becomes much harder if you have to figure out how many events are sitting in the PIC queue.

If I can figure out a good way to get SoundBlaster DMA working on its own thread, that would better model how it works on real hardware.  Again, haven't done a lot of research yet but I think I'll find on real hardware, the SoundBlaster doesn't use the PIC at all for this (instead the SoundBlaster reads asynchronously from memory while the CPU running the game writes to it).

Reply:
> Performance is a concern but also the ability to debug becomes much harder if you have to figure out how many events are sitting in the PIC queue.

The good news is there aren't a pile per-frame events in the PIC queue's linked list all sequenced out in time (so it's actually no different than the per-tick callback scenario).

When a callback is added to `TIMER_AddTickHandler(callback_function)`, the PIC system runs that function every tick, forever more, until it's explicitly removed from the handler.

Where when you use `PIC_AddEvent(callback_function, upcoming_event_millisecond)` that callback is only run one time (at the scheduled upcoming time), and then it's dropped from the PIC linked list.

So on both the per-frame and per-tick callback scenarios, there's only ever one upcoming callback in the PIC system to generate the next chunk of frame(s).

> If I can figure out a good way to get SoundBlaster DMA working on its own thread, that would better model how it works on real hardware. Again, haven't done a lot of research yet but I think I'll find on real hardware, the SoundBlaster doesn't use the PIC at all for this (instead the SoundBlaster reads asynchronously from memory while the CPU running the game writes to it).

Yup, that's exactly how DMA works. There's not PIC system involved; the device that uses the DMA memory makes hardware calls directly across the ISA bus and reads those memory locations without any concern for the CPU runtime state. 

When the game sets up the DMA transfer, is tells the DMA controller the main memory address it's writing data into, followed by how many elements its writing ("elements" being a generic quantity based on if the DMA channel is 16-bit or 8-bit). 

Ideally we could then copy that DMA transfer "object" into a thread safe space where the sound blaster could read from it fully async from the main thread, and paced out in time at the playback rate.